File: Emulator/src/main/core/game/component/Component.java
package core.game.component;

import core.api.Container;
import core.game.container.Slot;
import core.game.interaction.InterfaceListeners;
import core.game.node.entity.player.Player;
import core.game.node.entity.player.link.InterfaceManager;
import core.net.packet.PacketRepository;
import core.net.packet.context.InterfaceContext;
import core.net.packet.out.Interface;

import java.util.ArrayList;
import java.util.List;

/**
 * The type Component.
 */
public class Component {

    /**
     * The Id.
     */
    protected int id;

    /**
     * The Child id.
     */
    protected int childId = -1;

    private final List<Slot> slots = new ArrayList<>();

    /**
     * The Definition.
     */
    protected final ComponentDefinition definition;

    /**
     * The Close event.
     */
    protected CloseEvent closeEvent;

    /**
     * The Plugin.
     */
    protected ComponentPlugin plugin;

    private boolean hidden;

    /**
     * Instantiates a new Component.
     *
     * @param id the id
     */
    public Component(int id) {
        this.id = id;
        this.definition = ComponentDefinition.forId(id);
        this.plugin = definition != null ? definition.getPlugin() : null;
    }

    /**
     * Instantiates a new Component.
     *
     * @param id      the id
     * @param childId the child id
     */
    public Component(int id, int childId) {
        this.id = id;
        this.childId = childId;
        this.definition = ComponentDefinition.forId(id);
        this.plugin = definition != null ? definition.getPlugin() : null;
    }

    /**
     * Open.
     *
     * @param player the player
     */
    public void open(Player player) {
        InterfaceManager manager = player.getInterfaceManager();
        InterfaceListeners.runOpen(player, this);

        if (definition == null) {
            PacketRepository.send(Interface.class, new InterfaceContext(player, manager.getWindowPaneId(), manager.getDefaultChildId(), getId(), false));
            if (plugin != null) {
                plugin.open(player, this);
            }
            return;
        }

        if (definition.getType() == InterfaceType.WINDOW_PANE) {
            return;
        }

        if (definition.getType() == InterfaceType.TAB) {
            PacketRepository.send(Interface.class, new InterfaceContext(player, definition.getWindowPaneId(manager.isResizable()), definition.getChildId(manager.isResizable()) + definition.getTabIndex(), getId(), definition.isWalkable()));
            if (plugin != null) {
                plugin.open(player, this);
            }
            return;
        }

        PacketRepository.send(Interface.class, new InterfaceContext(player, definition.getWindowPaneId(manager.isResizable()), definition.getChildId(manager.isResizable()), getId(), definition.isWalkable()));
        if (plugin != null) {
            plugin.open(player, this);
        }
    }

    /**
     * Close boolean.
     *
     * @param player the player
     * @return the boolean
     */
    public boolean close(Player player) {
        return (closeEvent == null || closeEvent.close(player, this)) && InterfaceListeners.runClose(player, this);
    }

    /**
     * Handle slot switch boolean.
     *
     * @param player     the player
     * @param sourceSlot the source slot
     * @param destSlot   the dest slot
     * @return the boolean
     */
    public boolean handleSlotSwitch(Player player, int sourceSlot, int destSlot) {
        return InterfaceListeners.runSlotSwitch(player, this, sourceSlot, destSlot);
    }

    /**
     * Add slot.
     *
     * @param slotId    the slot id
     * @param container the container
     */
    public void addSlot(int slotId, Container container) {
        slots.add(new Slot(slotId, container));
    }

    /**
     * Gets slots.
     *
     * @return the slots
     */
    public List<Slot> getSlots() {
        return slots;
    }

    /**
     * Gets id.
     *
     * @return the id
     */
    public int getId() {
        return id;
    }

    /**
     * Gets child id.
     *
     * @return the child id
     */
    public int getChildId() {
        return childId;
    }

    /**
     * Gets definition.
     *
     * @return the definition
     */
    public ComponentDefinition getDefinition() {
        return definition;
    }

    /**
     * Gets close event.
     *
     * @return the close event
     */
    public CloseEvent getCloseEvent() {
        return closeEvent;
    }

    /**
     * Sets close event.
     *
     * @param closeEvent the close event
     * @return the close event
     */
    public Component setCloseEvent(CloseEvent closeEvent) {
        this.closeEvent = closeEvent;
        return this;
    }

    /**
     * Sets unclosable.
     *
     * @param player    the player
     * @param component the component
     */
    public static void setUnclosable(Player player, Component component) {
        player.setAttribute("close_c_", true);
        component.setCloseEvent(new CloseEvent() {
            @Override
            public boolean close(Player player, Component component) {
                return !player.getAttribute("close_c_", false);
            }
        });
    }

    /**
     * Sets plugin.
     *
     * @param plugin the plugin
     */
    public void setPlugin(ComponentPlugin plugin) {
        this.plugin = plugin;
    }

    /**
     * Gets plugin.
     *
     * @return the plugin
     */
    public ComponentPlugin getPlugin() {
        if (plugin == null) {
            ComponentPlugin p = ComponentDefinition.forId(getId()).getPlugin();
            if ((plugin = p) != null) {
                return p;
            }
        }
        return plugin;
    }

    /**
     * Is hidden boolean.
     *
     * @return the boolean
     */
    public boolean isHidden() {
        return hidden;
    }

    /**
     * Sets hidden.
     *
     * @param hidden the hidden
     */
    public void setHidden(boolean hidden) {
        this.hidden = hidden;
    }
}

File: Emulator/src/main/core/game/component/Component.kt
package core.game.component

import core.api.Container
import core.game.container.Slot
import core.game.interaction.InterfaceListeners.runClose
import core.game.interaction.InterfaceListeners.runOpen
import core.game.interaction.InterfaceListeners.runSlotSwitch
import core.game.node.entity.player.Player
import core.net.packet.OutgoingContext
import core.net.packet.PacketRepository
import core.net.packet.out.Interface

/**
 * Represents a UI component that can be opened, closed, and interacted with by a player.
 */
open class Component {
    /**
     * The unique identifier of this component.
     */
    @JvmField var id: Int

    /**
     * The child identifier of this component.
     */
    @JvmField var childId: Int = -1

    private val slots: MutableList<Slot> = ArrayList()

    /**
     * The definition of this component.
     */
    @JvmField
    val definition: ComponentDefinition?

    /**
     * The event triggered when the component is closed.
     */
    var closeEvent: CloseEvent? = null

    /**
     * The associated plugin for this component.
     */
    var plugin: ComponentPlugin?

    /**
     * Whether this component is hidden.
     */
    var isHidden: Boolean = false

    /**
     * Creates a new component with the specified ID.
     *
     * @param id The unique identifier of the component.
     */
    constructor(id: Int) {
        this.id = id
        this.definition = ComponentDefinition.forId(id)
        this.plugin = definition?.plugin
    }

    /**
     * Creates a new component with the specified ID and child ID.
     *
     * @param id      The unique identifier of the component.
     * @param childId The child identifier of the component.
     */
    constructor(id: Int, childId: Int) {
        this.id = id
        this.childId = childId
        this.definition = ComponentDefinition.forId(id)
        this.plugin = definition?.plugin
    }

    /**
     * Opens this component for the specified player.
     *
     * @param player The player opening the component.
     */
    open fun open(player: Player) {
        val manager = player.interfaceManager
        runOpen(player, this)

        if (definition == null) {
            PacketRepository.send(
                Interface::class.java,
                OutgoingContext.InterfaceContext(
                    player,
                    manager.windowPaneId,
                    manager.defaultChildId,
                    id,
                    false,
                ),
            )
            plugin?.open(player, this)
            return
        }

        if (definition.type == InterfaceType.WINDOW_PANE) {
            return
        }

        if (definition.type == InterfaceType.TAB) {
            PacketRepository.send(
                Interface::class.java,
                OutgoingContext.InterfaceContext(
                    player,
                    definition.getWindowPaneId(manager.isResizable),
                    definition.getChildId(manager.isResizable) + definition.tabIndex,
                    id,
                    definition.isWalkable,
                ),
            )
            plugin?.open(player, this)
            return
        }

        PacketRepository.send(
            Interface::class.java,
            OutgoingContext.InterfaceContext(
                player,
                definition.getWindowPaneId(manager.isResizable),
                definition.getChildId(manager.isResizable),
                id,
                definition.isWalkable,
            ),
        )
        plugin?.open(player, this)
    }

    /**
     * Closes this component for the specified player.
     *
     * @param player The player closing the component.
     * @return `true` if the component was closed successfully, otherwise `false`.
     */
    open fun close(player: Player): Boolean = (closeEvent == null || closeEvent!!.close(player, this)) && runClose(player, this)

    /**
     * Handles slot switching within the component.
     *
     * @param player     The player interacting with the slots.
     * @param sourceSlot The source slot index.
     * @param destSlot   The destination slot index.
     * @return `true` if the slot switch was handled successfully, otherwise `false`..
     */
    fun handleSlotSwitch(
        player: Player,
        sourceSlot: Int,
        destSlot: Int,
    ): Boolean = runSlotSwitch(player, this, sourceSlot, destSlot)

    /**
     * Adds a new slot to this component.
     *
     * @param slotId    The slot identifier.
     * @param container The container associated with the slot.
     */
    fun addSlot(
        slotId: Int,
        container: Container,
    ) {
        slots.add(Slot(slotId, container))
    }

    /**
     * Retrieves the list of slots associated with this component.
     *
     * @return A list of slots.
     */
    fun getSlots(): List<Slot> = slots

    /**
     * Sets the close event for this component.
     *
     * @param closeEvent The close event to set.
     * @return This component instance.
     */
    fun setUncloseEvent(closeEvent: CloseEvent?): Component {
        this.closeEvent = closeEvent
        return this
    }

    companion object {
        /**
         * Marks a component as unclosable.
         *
         * @param player    The player for whom the component should be unclosable.
         * @param component The component to mark as unclosable.
         */
        fun setUnclosable(
            player: Player,
            component: Component,
        ) {
            player.setAttribute("close_c_", true)
            component.setUncloseEvent { player, _ ->
                !player.getAttribute("close_c_", false)
            }
        }
    }
}
